common/materialContainer/MaterialContainer.cpp

```diff
@@ -123,83 +116,153 @@ Ground Ground::noOre()
   vec.push_back(Material::Sand());
 
   std::string name = "noOre";
-  Ground g(name, vec, .5, 10.0);
-  return g;
+  return std::make_shared<GroundContainer>(name, vec, .5, 10.0);
 }
 
-Ground Ground::soil()
+std::shared_ptr<GroundContainer> GroundContainer::soil()
 {
-  std::vector<Material> vec;
+  std::vector<std::shared_ptr<Material>> vec;
   vec.push_back(Material::Soil());
 
   std::string name = "soil";
-  Ground g(name, vec, .5, 10.0);
-  return g;
+  return std::make_shared<GroundContainer>(name, vec, .5, 10.0);
 }
 
-Ground Ground::rockySoil()
+std::shared_ptr<GroundContainer> GroundContainer::rockySoil()
 {
-  std::vector<Material> vec;
+  std::vector<std::shared_ptr<Material>> vec;
   vec.push_back(Material::Stone());
   vec.push_back(Material::Soil());
   vec.push_back(Material::Gravel());
 
   std::string name = "rockySoil";
-  Ground g(name, vec, .2, 10.0);
-  return g;
+  return std::make_shared<GroundContainer>(name, vec, .2, 10.0);
 }
 
-Ground Ground::sandySoil()
+std::shared_ptr<GroundContainer> GroundContainer::sandySoil()
 {
-  std::vector<Material> vec;
+  std::vector<std::shared_ptr<Material>> vec;
   vec.push_back(Material::Sand());
   vec.push_back(Material::Soil());
 
   std::string name = "sandySoil";
-  Ground g(name, vec, .2, 10.0);
-  return g;
+  return std::make_shared<GroundContainer>(name, vec, .2, 10.0);
 }
 
-Material Ground::getMaterial()
+std::shared_ptr<Material> GroundContainer::getMaterial(int x, int y, int z)
 {
-  double noise = PerlinNoise();
+  double noise = PN(x, y, z);
   return getElement(noise);
 }
 
-Material Material::Coal()
+std::shared_ptr<Material> Material::makenew(std::string name, BlockType type, double freq){
+  materialMap.insert(std::pair<std::string,std::shared_ptr<Material>>(name,std::make_shared<Material>(name,type,freq)));
+  return materialMap[name];
+}
+
+std::shared_ptr<Material> Material::getFromType(BlockType BT)
+{
+  switch(BT)
+  {
+    case BlockType::Coal:
+       return Material::Coal(); 
+     
+     case BlockType::Iron:
+       return Material::Iron(); 
+
+     case BlockType::Ground:
+       return Material::Soil(); 
+
+     case BlockType::Gravel:
+       return Material::Gravel(); 
+
+     case BlockType::Sand:
+       return Material::Sand(); 
+
+     case BlockType::Diamond:
+       return Material::Diamond(); 
+
+     case BlockType::Stone:
+       return Material::Stone(); 
+
+     case BlockType::Gold:
+       return Material::Gold();
+
+     default:
+       return Material::Default(); 
+  }
+}
+
```

Comment:
Now that I see the case branching on `BlockType` in practice, I think it could get messy. 

Is `MaterialContainer` supposed to be implement a flyweight pattern for `Material`? That is, does `MaterialContainer` *only* memoize all calls when a caller tries to get a material, or does it do something more complicated?

If it's the former, you could try flyweighting the Block class. Each block could share a map from `BlockType` to `Material`. Might be useful if you need more information per block instance than just its type.